src/Microsoft.AspNetCore.Routing/Builder/EndpointRoutingApplicationBuilderExtensions.cs

```diff
@@ -18,20 +18,39 @@ public static IApplicationBuilder UseEndpointRouting(this IApplicationBuilder bu
             return builder.UseEndpointRouting(null);
         }
 
-        public static IApplicationBuilder UseEndpointRouting(this IApplicationBuilder builder, Action<EndpointDataSourceBuilder> configure)
+        public static IApplicationBuilder UseEndpointRouting(this IApplicationBuilder builder, Action<IEndpointRouteBuilder> configure)
         {
             VerifyRoutingIsRegistered(builder);
 
+            var globalEndpointDataSource = builder.ApplicationServices.GetRequiredService<CompositeEndpointDataSource>();
+            CompositeEndpointDataSource middlewareEndpointDataSource;
+
             if (configure != null)
             {
-                var dataSourceBuilder = (DefaultEndpointDataSourceBuilder)builder.ApplicationServices.GetRequiredService<EndpointDataSourceBuilder>();
-                dataSourceBuilder.ApplicationBuilder = builder;
-                configure(dataSourceBuilder);
+                middlewareEndpointDataSource = new CompositeEndpointDataSource();
+
+                var endpointRouteBuilder = builder.ApplicationServices.GetRequiredService<IEndpointRouteBuilder>();
+                if (endpointRouteBuilder is DefaultEndpointRouteBuilder defaultEndpointRouteBuilder)
+                {
+                    defaultEndpointRouteBuilder.ApplicationBuilder = builder;
+                }
+                configure(endpointRouteBuilder);
+
+                foreach (var dataSource in endpointRouteBuilder.DataSources)
+                {
+                    // REVIEW: Should there be a way to add multiple datasources at once to avoid potentially re-triggering change token multiple times?
```

Comment:
Well the way to do that would be to have an 'AddRange` method. I don't really see it as a problem because I don't really think anyone is listening to the change tokens yet. Everything gets initialized on the first request (or first link generation).